--- cvs-fast-export-1.66/treepack.c.orig	2024-04-07 16:00:03.669621200 +0200
+++ cvs-fast-export-1.66/treepack.c	2024-04-07 16:01:11.059256119 +0200
@@ -210,6 +210,7 @@ void revdir_pack_init(void) {
 	frame = frames;
 	nfiles = 0;
 	frames[0].dir = root_dir;
+	frames[0].dirs = xmalloc(0, __func__);
 	frames[0].ndirs = 0;
 	frames[0].hash = hash_init();
 }
@@ -253,6 +254,7 @@ void revdir_pack_add(const cvs_commit *f
 
 			const master_dir *parent = frame++->dir;
 			frame->dir = first_subdir(dir, parent);
+			frame->dirs = xmalloc(0, __func__);
 			frame->ndirs = 0;
 			frame->hash = hash_init();
 			continue;
--- cvs-fast-export-1.66/generate.c.orig	2024-04-07 16:00:03.669621200 +0200
+++ cvs-fast-export-1.66/generate.c	2024-04-07 16:02:04.075635571 +0200
@@ -1158,7 +1158,7 @@ static node_t *generate_setup(generator_
 		eb->Gfilename = gen->master_name;
 		eb->Gexpand = gen->expand;
 		eb->Gabspath = NULL;
-		Gline(eb) = NULL;
+		Gline(eb) = xmalloc(0, __func__);
 		Ggap(eb) = Ggapsize(eb) = Glinemax(eb) = 0;
 	}
 
--- cvs-fast-export-1.66/import.c.orig	2024-04-07 16:00:03.669621200 +0200
+++ cvs-fast-export-1.66/import.c	2024-04-07 16:02:53.108703269 +0200
@@ -347,6 +347,7 @@ void analyze_masters(int argc, const cha
 	striplen = analyzer->striplen;
 
 	forest->textsize = forest->filecount = 0;
+	forest->is_cvs = false;
 	progress_begin("Reading file list...", NO_MAX);
 	for (;;) {
 		struct stat stb;
