--- cvs-fast-export-1.62/export.c.orig	2023-12-10 22:08:38.039921414 +0100
+++ cvs-fast-export-1.62/export.c	2023-12-11 06:17:08.424656665 +0100
@@ -208,7 +208,7 @@ static void export_blob(node_t *node,
 	for (char *cp = cbuf; cp < cbuf + len; cp++)
 	    if (*cp == ' ')
 		*cp = '\n';
-	if (strlen(cbuf) >= 2 && cbuf[0] == '!' && cbuf[1] == '\n')
+	if (strnlen(cbuf, len) >= 2 && cbuf[0] == '!' && cbuf[1] == '\n')
 	    extralen = 0;
     }
 
@@ -697,7 +697,7 @@ static struct commit_seq *canonicalize(g
 #define is_branchroot_of(x, y) ((x)->head == (y)->head && (x)->isbase)
 #define is_older_than(x, y) (((struct commit_seq *)x)->commit->date < ((struct commit_seq *)y)->commit->date)
 	/* back up as far as we can */
-	while (!is_parent_of(bp-1, hp) && !is_branchroot_of(bp-1, hp) && !is_older_than(bp-1, hp))
+	while ((bp > history) && !is_parent_of(bp-1, hp) && !is_branchroot_of(bp-1, hp) && !is_older_than(bp-1, hp))
 	    bp--;
 	if (bp < hp) {
 	    /* shift commits up and put *hp where *bp was */
